### ISSUE TITLE
Users thought key was a required parameter

### ISSUE THREAD

**Issue by InMatrix** [flutter/flutter#10219]
## Steps to Reproduce

Key is the first item in the parameter list of a widget, while it's not commonly needed. The last UX study participant initially thought key was required to create a Card, since key was the first thing he saw in the signature the IDE showed him, until he inspected the source of Card. 

![image](https://cloud.githubusercontent.com/assets/348942/26269465/83243ab8-3ca9-11e7-9bdc-2953716b73d7.png)

Can we consider moving key to the last of the parameter list? It's helpful to have more commonly used parameters come first. It should not affect API stability since key is a named parameter. 

If changing the code is not a good idea, is it possible to change the presentation of constructor parameters in the IDE?

## Flutter Doctor
```
[✓] Flutter (on Mac OS X 10.12.4 16E195, channel master)
    • Flutter at /Users/taodong/Code/flutter_repos/flutter
    • Framework revision a88379399e (3 hours ago), 2017-05-19 12:34:41 -0700
    • Engine revision 82815836eb
    • Tools Dart version 1.23.0-dev.11.11

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.2)
    • Android SDK at /Users/taodong/Library/Android/sdk
    • Platform android-24, build-tools 25.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version: OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 8.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.2.1, Build version 8C1002
    • ios-deploy 1.9.1
    • CocoaPods version 1.2.0

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Gradle version 3.2
    • Java version: OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Community Edition (version 2017.1)
    • Dart plugin version 171.4006
    • Flutter plugin version 13.1
```

**Comment by InMatrix** [comment]
cc: @Hixie @pq

**Comment by eseidelGoogle** [comment]
Maybe more generally flutter-intellij should show @required parameters first?  We certainly also can do manual ordering based on likely-hood of use. :) . FYI @devoncarew

**Comment by sethladd** [comment]
Good catch. Thanks for bringing this up!

**Comment by devoncarew** [comment]
Manual ordering does change the order that IntelliJ displays the params in, so probably a good thing to do. Definitely possible to sort based on `@required` as well. Perhaps these should be listed first in the definition anyway?

**Comment by pq** [comment]
👍 

We could definitely sort `@required` first.  That said, is there any reason not to do that in the definition of `Card`?  Might be generally good style to recommend putting common (and especially required) parameters first in declarations (@munificent?).

**Comment by Timmmm** [comment]
Just to add to this, I also thought `key` was required. Not just because it is first, but because in [the constructor signature](https://docs.flutter.io/flutter/widgets/Wrap/Wrap.html) it doesn't seem to have a default value like the other parameters. I guess Dart has a default implicit value of `null` or something, but I'm new to Dart (as I imagine most Flutter users are) and didn't know that and nothing in the docs I found says that you can leave it as `null`.

I was trying to use the `Wrap` layout, and first looked there for an example to see what `key` should be set to but there isn't one. Then I looked in [the docs for `Key`](https://docs.flutter.io/flutter/widgets/Key-class.html) but they don't really explain anything, other than that it is a unique key for widgets which I kind of guessed. Finally I realised that none of the other widgets take a `key` in the examples so maybe I can just leave it out. Tried it and it worked.

So to improve this you can:

1. Add basic examples to the API reference for all widgets/layouts showing at least how to create one (without using key).
2. Improve the `Key` documentation to explain what it actually is and that you can normally leave it as `null`.

**Comment by pq** [comment]
cc @lukechurch who may have thoughts.

Great feedback @Timmmm.  Thank you!

**Comment by Hixie** [comment]
@Timmmm for #2, can you clarify which of the following you looked at?
- https://docs.flutter.io/flutter/widgets/Key-class.html 
- https://docs.flutter.io/flutter/widgets/Widget/key.html
- https://docs.flutter.io/flutter/widgets/Wrap/Wrap.html
- https://docs.flutter.io/flutter/widgets/Wrap-class.html
- something else

For #1, I have a patch out which adds some sample code for Wrap:
https://github.com/flutter/flutter/pull/10214/files#diff-7f71460835520a1820d69cd0b6994c01R2396
It's just awaiting review before I can land it.

**Comment by sethladd** [comment]
The request for more docs re: Keys is at https://github.com/flutter/flutter/issues/8818

**Comment by Timmmm** [comment]
@Hixie 

I looked at all of those except https://docs.flutter.io/flutter/widgets/Widget/key.html ... which apparently has the best explanation!

Good work on the example. :-)

**Comment by Hixie** [comment]
Cool, thanks. I'll make sure to make Key-class point to that. Taking this to #8818.

**Comment by munificent** [comment]
> Might be generally good style to recommend putting common (and especially required) parameters first in declarations (@munificent?).

The language requires mandatory parameters appear before optional ones, so the style guide doesn't need to. (Neither the style guide nor the language know about `@required`, and the style guide generally avoids mentioning stuff that's not part of the language or core libraries.)

It's worded a little funny, but Effective Dart [does say](https://www.dartlang.org/guides/language/effective-dart/design#avoid-optional-positional-parameters-if-the-user-may-want-to-omit-earlier-parameters) the most commonly passed optional *positional* parameters should appear earlier in the parameter list.

It doesn't say anything about named parameters since it doesn't matter as much with them, but maybe it should. Would you mind filing a tracking bug for that: https://github.com/dart-lang/site-www

I don't have much time for Effective Dart, but I'm hoping to do a big update before too long.

**Comment by sethladd** [comment]
Seems reasonable to move key to the end (or near end) of the parameter list.

Consider this another +1 for moving @required into formal language semantics. Required named parameters pretty please :)

**Comment by Hixie** [comment]
In Flutter we like to have the parameters be listed in the same order as the fields and the same order as the arguments. Since we normally want `child` to be the last argument, we like to have `child` be the last parameter and the last field. (We're not very consistent about this for historical reasons, but that's the goal.)

The `child` argument is often required. This is one reason why we often have `@required` arguments after optional ones.

**Comment by sethladd** [comment]
> In Flutter we like to have the parameters be listed in the same order as the fields

Why does the user care about this?

**Comment by Hixie** [comment]
It makes the code easier to read. In particular, it makes it easier to compare arguments, parameters, and fields so that you can tell if you've included all the ones you want to include. For example, we also try to have the same order in `hashCode` and `operator ==` implementations. Having the same order makes it much more obvious when you've missed one (just earlier today, I found a bug that would have been easier to find if we hadn't had the same list of identifiers in four different orders in the same file...).

Anyway, long story short, users care because fewer bugs means a better user experience. :-)

**Comment by sethladd** [comment]
So what's the suggestion for code completion to not show Key as the first parameter?

**Comment by Hixie** [comment]
I agree that IDEs would serve the users well by having the @required named arguments first. I would probably have the @required arguments first, alphabetically, followed by the others, alphabetically.

**Comment by pq** [comment]
Great feedback!  I'm curious if it would be surprising if the order was different than defined in the source.  (I'm thinking of feedback regarding adaptive menus.)  Either way, I'm curious why this isn't done in the source.  Wouldn't it benefit documentation if required params came first?  Or is there another rationale in how optional named params are ordered?

Relatedly, I see that the source part of the API docs has the required annotations 

<img width="629" alt="screen shot 2017-05-23 at 6 38 18 am" src="https://cloud.githubusercontent.com/assets/67586/26356887/713e9fec-3f82-11e7-8995-fc1b25ab7afb.png">

but not the summary at the top

<img width="196" alt="screen shot 2017-05-23 at 6 36 17 am" src="https://cloud.githubusercontent.com/assets/67586/26356827/3adc5a34-3f82-11e7-8b97-627dad12e8c6.png">

Are there plans to make the required bits pop out a bit more? cc @jcollins-g

**Comment by pq** [comment]
Looking back at @munificent's comment:

> It's worded a little funny, but Effective Dart does say the most commonly passed optional positional parameters should appear earlier in the parameter list.

This seems to add a little more fuel to the sort required params first in source proposal?

On the other hand, I see @Hixie wrote:

> In Flutter we like to have the parameters be listed in the same order as the fields and the same order as the arguments. 

Which I get from an implementation perspective but not so much for documentation (but maybe they're orthogonal).

@InMatrix @lukechurch : any more thoughts from a usability perspective?

**Comment by jcollins-g** [comment]
dart-lang/dartdoc#1432 seems relevant to this.

**Comment by InMatrix** [comment]
Yes, dart-lang/dartdoc#1432 is relevant here. IMO, we should fix that issue first, and then also make sure required parameters are consistently presented in IntelliJ. Right now, they are not. 

Let's take IconButton as an example:

As I type "new IconButton", I won't see what's required in the autocomplete box:
![image](https://cloud.githubusercontent.com/assets/348942/26364150/30dfacaa-3f98-11e7-987c-c0da112c9d11.png)

If I'm lucky to hit tab to complete the signature, the IDE does cleverly populate the two required parameters for me. But I appreciate this because I already know these two parameters are required. To a new user, it might not be clear why these two parameters are more important than others, given the tooltip below showing all the params without indicating these two are required. 

![image](https://cloud.githubusercontent.com/assets/348942/26364183/5091f0da-3f98-11e7-9144-2d1e440d77b8.png)

If I didn't hit tab but finished typing "IconButton()", I can manually open the autocomplete box but then the list doesn't show me what's required, nor default/inherited values if they existed.
![image](https://cloud.githubusercontent.com/assets/348942/26364134/24b6f870-3f98-11e7-9698-dce4a49eb6c7.png)

I also agree with the @Timmmm's comment. It's often hard to determine whether a parameter has a default value. That requires a separate fix. But making required parameters more prominent and hard to miss will help users calm down when they deal with any non-required parameters. 

Re: adaptive menus 

The issues with adaptive menus are a little different from the potential inconsistency between the order of params in source and the order of those in the IDE. Adaptive menus change the order of menu items dynamically based on the frequency of use at runtime for each individual. In contrast, the re-ordering proposal we're discussing here will be done at design time and will be the same for all the users. 

Given most widgets seem to have no more than a couple of required parameters (correct?), I would say the benefit is likely to outweigh the impact of inconsistency with the order in source. That said, @Hixie's rationale against changing parameter order in source seems to be mostly relevant to framework developers and advanced users who need to write new Flutter widgets, no?

**Comment by Hixie** [comment]
As I said earlier:

> [In IDEs] I would probably have the @required arguments first, alphabetically, followed by the others, alphabetically.

**Comment by zoechi** [comment]
seems related dart-lang/linter#678

**Comment by Hixie** [comment]
cc @mit-mit Moving this to m3 since "editing build methods is easy" is on our radar for m3 and this falls into that category.

**Comment by mit-mit** [comment]
cc @devoncarew where (analyzer vs tools vs dart plugin vs flutter plugin) would this change need to be made? It would be nice if we could make it work in all IDEs based on the analyzer.

**Comment by pq** [comment]
FWIW a hacky stab at fixing this is in flight: https://dart-review.googlesource.com/c/sdk/+/38346.

Here's how it affects creating `LayoutId`s:

![image](https://user-images.githubusercontent.com/67586/35758304-2f722790-0828-11e8-8bec-3ad29e5df0d4.png)

(Note that key gets sorted last. 🕺 )

**Comment by pq** [comment]
Landed in https://github.com/dart-lang/sdk/commit/1399a5f511907f63b104f78257af7982b7fe8dfc 👍

**Comment by Timmmm** [comment]
Nice work - I'm really impressed how much effort you guys are putting into making Flutter easy to use. Makes a nice change from Android.

**Comment by sethladd** [comment]
Thanks @pq !

BTW, how does this work? How does it know to suggest Key at the end?

**Comment by pq** [comment]
@sethladd : it's really simple actually.  We're following @eseidelGoogle's suggestion and simply sorting required params first; a happy result is that `key` being truly optional lands at the end.

**Comment by pq** [comment]
Thanks for the kind words @Timmmm!

**Comment by sethladd** [comment]
@pq can we visually show (somehow, maybe with actually the literal `@required`) the required-ness of the parameters, in the code-complete pop-up?

it's not obvious which are required, and which are optional.

**Comment by sethladd** [comment]
Asking another way: how will our developers know what is required and what is optional? Is there another UI hint somewhere?

(I believe our challenge is to clearly indicate what is _required_ when code completing. We think that's important because if you finish code completing a constructor, and then you get warnings, you're unhappy and now you have to go back and resolve the warning. Ideally, the journey of completing a constructor happens once, as you are adding at least required parameters.)

**Comment by bwilkerson** [comment]
We provide a hint when a required parameter is missing.

We sort required parameters first when code completing inside an argument list.

We've talked about trying to do something in the completion list such as bolding the required parameters, but I don't know whether that's possible or whether we've looked into it.

**Comment by Hixie** [comment]
Can you elaborate on what more will need to be done here, if anything, for us to consider this fixed?

**Comment by pq** [comment]
@Hixie : I guess all that remains is a decision on whether to add a `@required` prefix to required params in the popup or whether sorting them first is sufficient.

**Comment by InMatrix** [comment]
@pq: As an alternative to displaying `@required`, can we reduce the contrast of optional parameters? 

In addition, I think we can display `@required` when the autocomplete box is invoked within an expanded constructor:
![image](https://user-images.githubusercontent.com/348942/36234219-a985af14-119e-11e8-8492-b81eaa2192f2.png)

There is enough real estate in this case.

**Comment by pq** [comment]
I'm not sure about changing the style within the completion though I love the idea.  I was imagining bolding required params would be nice.  So, something like:

 Thing({**Widget child**, Color color});

where "child" is required and "color" is not.

I know it's easy to change the style of the entire lookup text but I'm not sure we can have different styles within it.

cc @alexander-doroshko for thoughts

**Comment by alexander-doroshko** [comment]
`LookupElementBuilder` has a bunch of standard decorations: bold, strikeout, underlined, tail text, type text, icon, foreground color. If it's not enough there're a few more options available if you use `LookupElementBuilder.withRenderer()`.

**Comment by pq** [comment]
Cool.  Thanks @alexander-doroshko!

Looking ahead, 

1. adding a `@required` prefix could happen quickly (as it's all in server)
2. custom rendering would take longer (and is on the IDEA side)

If folks elect to go with (1), I can knock that out pronto.  Otherwise, I think we should open a feature request on the Dart Plugin for (2).  Alternatively we can stick w/ (0) and leave it as is.

**Comment by sethladd** [comment]
@pq would love a mockup of adding `@required`, thank you!

(I'll take you up on the offer for "could happen quickly" :)

**Comment by pq** [comment]
😄 

Quick mock:

![feb-15-2018 12-51-53](https://user-images.githubusercontent.com/67586/36280749-28db1e46-1250-11e8-9035-4d0ab0f4f773.gif)

@sethladd : Happy to run on "real" Flutter classes.  Just tell me what you'd like to see!

**Comment by sethladd** [comment]
Whoa, bonus points for animated gif!!

Works for me. Now it's super clear what I need to do to have a correct-on-the-first-try constructor call.

Later, when we figure out how to control colors/fonts in there, we can play more with the aesthetics.

I'd say SHIP IT!

**Comment by pq** [comment]
Groovy.

Landed in: https://github.com/dart-lang/sdk/commit/3dd7eae4c8d0f77aa76b6222d039d473b8bf8261

@DanTup, @alexander-doroshko : this should "just work" out of the box but please let me know if you see any issues.

We can revisit, down the road, if/when we want to refine rendering.

**Comment by sethladd** [comment]
Thanks for the quick turnaround!

**Comment by DanTup** [comment]
@pq Where are those args in the completion coming from? Are they being added to the new `displayText` field? Currently in Code we don't display arguments in completion labels (we just do `(...)`), however since we append that on to the end of the label from the server, if this is new text being added into `displayText` (or even `completion`) then we might end up appending further parens/args to the end. I tried looking at the CLs above but couldn't figured out exactly what the change is.

**Comment by pq** [comment]
@DanTup : the args are in the `parameters` property of the `Element` coming back from server. Looking at how it's done in IntelliJ will probably help...  See [DartServerCompletionContributor.java#L291](https://github.com/JetBrains/intellij-plugins/blob/master/Dart/src/com/jetbrains/lang/dart/ide/completion/DartServerCompletionContributor.java#L291):

    final String parameters = element.getParameters();

(and beyond).

This particular fix doesn't involve `displayText` at all.  (Though it probably should!)

**Comment by DanTup** [comment]
Aha, I was confused because I remembered using an array. Looks like we only use that string in the `detail` (additional info the user can expand, not the completion text nor the display text) and we parse from parameterNames for the completion text:

<img width="1014" alt="screen shot 2018-02-18 at 17 06 15" src="https://user-images.githubusercontent.com/1078012/36354551-3a3697f8-14ce-11e8-950d-6fadd7ce9965.png">

So, assuming the order of `parameterNames` hasn't changed, I don't think it'll make any difference to Dart Code (except the ordering in the additional info).

> This particular fix doesn't involve displayText at all. (Though it probably should!)

I think now we have `displayText` it'd be good to review the completion/displayText for existing items - Dart Code has a bunch of custom logic here that I'd love to just ditch and show raw values from the server (we'd have to agree on what's best, but I'm we can manage that). That said, because of all of that custom logic (above) we'd need to coordinate changing it (probably just a bump in the analysis server version would be good enough for me) to avoid displaying weird things to users as the change rolls out.

**Comment by pq** [comment]
@DanTup 👍 Absolutely.  Doing work on the server side and populating `displayText` appropriately is very much the plan.  (FWIW there's similar logic on the IntelliJ side and it would be great to make the cause for it to go away as well with one fell swoop.)

**Comment by DanTup** [comment]
@pq I just reviewed my code, and it turns out there are only two minor tweaks we make to display text, most of the logic is the snippet handling and the "detail" (a bigger info window shown for the selected item).

The display text tweaks we make are:

- Append `()` or `(…)` depending on whether there are parameters (except for getters, where we append neither)
- Remove any trailing comma (flutter formatting triggers)

I think the bottom one probably makes sense in the server, but the top one I'm not sure - if other editors are adding `parameters` onto the end themselves (we don't, because we didn't want a really long line, hence the `(…)`) then this would break that.

**Comment by pq** [comment]
Interesting.  We'll have to review what we do in IntelliJ as well.  Sounds like this isn't a huge deal either way though.  Thanks for following up!
